### PR TITLE
Assignment4 -Victor Korn

### PR DESCRIPTION
**Pull request by vicdotexe**

### CODE REVIEW

**Review by randypipkins — COMMENTED** [review]
All code compiles and runs successfully and all unit tests succeed. Only suggestion I might make would be to make sure your { } are on separate lines when declaring get; and set; to improve code readability

File: Logger.Tests/TestEntity.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        var last = "Korn";
+        var middle = "Ernest";
+        var fullName = new FullName(first, last, middle);
+        var person = new Person(fullName);
+        Assert.AreEqual(person.Name, $"{first} {middle} {last}");
+    }
+
+    [TestMethod]
+    public void StudentTest()
+    {
```
Comment (mmoran314) [inline Logger.Tests/TestEntity.cs:64]: Maybe add tests for the added parameters for each record, like major in student.

File: Logger/Employee.cs [hunk 2]
```diff
@@ -0,0 +1,6 @@
+﻿namespace Logger;
+
+public record class Employee(FullName FullName, string Role) : Person(FullName)
+{
```
Comment (mmoran314) [inline Logger/Employee.cs:4]: I think that these classes need properties for the added parameters i.e. for Role in the Employee class and Major in student class

**Review by mmoran314 — COMMENTED** [review]
Define a full name record (first, last, middle) handling optional and null appropriately. ✔
Provide a comment on the full name record on why you selected to define a value or a reference type and ✔
Provide a comment on the full name record on why or why not the type is immutable. ✔
Define an IEntity interface: ✔
Add an Id property of type Guid that is an init-only setter. ✔
Add a Name property that is string. ✔
Define an abstract base class that implements IEntity - appropriately choosing to implement the interface explicitly or implicitly. ✔
Do not implement the Name property in this abstract class. ✔
Do force any derived classes to provide an implementation for Name. ✔
Define book, student, and employee records - all with entity capabilities. ✔
Choose a thoughtful approach for the value of Name for each type of entity. In all cases, there should be no backing field for Name (not even one generated by an automatically implemented property) - consider a calculated property. ✔
Provide a comment on each interface member in each entity explaining why you implemented it implicitly or explicitly. ✔
You should consider the relationship between Student and Employee and refactor the common code shared between them. ✔
Given the properties on the full name record, you should consider which entities it makes sense to use it. ✔
Test that the equality behavior on these entities behaves as expected. ✔
Provide a comment on each interface method explaining why you implemented it implicitly or explicitly.
Unit test the Storage class using various entities to ensure that there are no bugs.
Code review (using TODO: comments if there are any issues - which there may not be) all logger-related classes.)
Pay special attention to the CreateLogger methods throughout.
Fundamentals
Be sure you enable:
Enable Nullability for all projects ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Ensure there are no errors or warnings (including code analysis warnings) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

I think it looks good. I think that you might need to add to some of the records for the new parameters that they have so that everything gets implemented when creating the new instance of the object, adding tests for these as well.

**Comment by mmwoodfo** [comment]
> All code compiles and runs successfully and all unit tests succeed. Only suggestion I might make would be to make sure your { } are on separate lines when declaring get; and set; to improve code readability

If the getter and setters are just being declared, they should stay inline. Is there a specific area in the code you were referring to here?

File: Logger/Employee.cs (hunk 2 above)
Comment (vicdotexe) [reply Logger/Employee.cs:4]: Like I need to override the ones that a record automatically generates? I don't understand why.

File: Logger/FullName.cs [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+//and safeguard against assuming the changes have affected any copies.
+public readonly record struct FullName(string FirstName, string LastName, string? MiddleName = null)
+{
+    public string FirstName { get; } = FirstName ?? throw new ArgumentNullException(nameof(FirstName));
+    public string LastName { get; } = LastName ?? throw new ArgumentNullException(nameof(LastName));
+    public string? MiddleName { get; } = MiddleName;
+
+    public override string ToString()
+    {
+        return MiddleName == null ? $"{FirstName} {LastName}" : $"{FirstName} {MiddleName} {LastName}";
```
Comment (mmwoodfo) [inline Logger/FullName.cs:18]: Nice, you can also use something called pattern matching here for a more natural feel when reading this statement (note, this pattern matching only works for equality when checking enums and null values)
```suggestion
        return MiddleName is null ? $"{FirstName} {LastName}" : $"{FirstName} {MiddleName} {LastName}";
```

File: Logger/IEntity.cs [hunk 4]
```diff
@@ -1,7 +1,6 @@
 ﻿namespace Logger;
 public interface IEntity
 {
-
-        // Place members here.
-
+    Guid Id { get; init; }
+    string? Name { get; }
```
Comment (mmwoodfo) [inline Logger/IEntity.cs:5]: I don't think you need to make name nullable here.

File: Logger/Employee.cs (hunk 2 above)
Comment (mmwoodfo) [reply Logger/Employee.cs:4]: You do not need to declare parameters unless you intend on implementing any kind of getter/setter logic on the properties.

File: Logger.Tests/TestEntity.cs [hunk 5]
```diff
@@ -0,0 +1,70 @@
+﻿using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Logger.Tests;
+
+[TestClass]
+public class TestEntity
+{
+    [TestMethod]
+    public void Book_CompareDifferentIds_NotEqual()
```
Comment (mmwoodfo) [inline Logger.Tests/TestEntity.cs:9]: Good job adding inequality checks as well!

**Pull request closed without merging**